vowpalwabbit/loss_functions.cc

```diff
@@ -292,6 +322,57 @@ class quantileloss : public loss_function
   float tau;
 };
 
+// Expectile loss is closely related to the squared loss, but it's an asymmetric function with a quantile parameter
+// Its methods can be derived from the corresponding methods from the squared loss multiplied by the quantile value
+class expectileloss : public loss_function
+{
+public:
+  expectileloss(float& q_) : q(q_) {}
+
+  std::string getType() const override { return "expectile"; }
+  float getParameter() const override { return q; }
+
+  float getLoss(const shared_data* sd, float prediction, float label) const override
+  {
+    float err = label - prediction;
+    return squared_loss_impl::getLoss(sd, prediction, label) * (err > 0 ? q : (1.f - q));
+  }
+
+  float getUpdate(float prediction, float label, float update_scale, float pred_per_update) const override
+  {
+    float err = label - prediction;
+    return err > 0 ? squared_loss_impl::getUpdate(prediction, label, q * update_scale, pred_per_update)
+                   : squared_loss_impl::getUpdate(prediction, label, (1.f - q) * update_scale, pred_per_update);
+  }
+
+  float getUnsafeUpdate(float prediction, float label, float update_scale) const override
+  {
+    float err = label - prediction;
+    return err > 0 ? squared_loss_impl::getUnsafeUpdate(prediction, label, q * update_scale)
+                   : squared_loss_impl::getUnsafeUpdate(prediction, label, (1.f - q) * update_scale);
+  }
+
+  float getSquareGrad(float prediction, float label) const override
+  {
+    float err = label - prediction;
+    return squared_loss_impl::getSquareGrad(prediction, label) * (err > 0 ? q * q : (1.f - q) * (1.f - q));
+  }
+
+  float first_derivative(const shared_data* sd, float prediction, float label) const override
+  {
+    float err = label - prediction;
+    return squared_loss_impl::first_derivative(sd, prediction, label) * (err > 0 ? q : (1.f - q));
+  }
+
+  float second_derivative(const shared_data* sd, float prediction, float label) const override
+  {
+    float err = label - prediction;
+    return squared_loss_impl::second_derivative(sd, prediction) * (err > 0 ? q : (1.f - q));
+  }
+
+  float q;
```

Comment:
Member variables should be prefixed with `_`
```suggestion
  float q;
```

Reply:
Please either give the variable a more descriptive name or add a comment explaining what `q` means for the benefit of future people reading this file

Reply:
Please also make `q` private.